File: LeetCode/Medium/0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.ts
function canPartition(nums: number[]): boolean {
    const sum = nums.reduce((a, b) => a + b, 0)
    if (sum % 2 !== 0) return false;
    const target = sum / 2

    const dp: number[] = new Array(target + 1).fill(0);

    for (const x of nums) {
        for (let j = target; j >= x; j--) {
            dp[j] = Math.max(dp[j], dp[j - x] + x);
        }
        // 可选早停：凑满了就直接返回
        if (dp[target] === target) return true;
    }
    return dp[target] === target;
};

File: LeetCode/Medium/0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {

        int sum = accumulate(nums.begin(), nums.end(), 0);

        if (sum % 2 != 0) return false;

        int target = sum / 2;

        vector<int> dp(10001, 0);

        for (int i = 0; i < nums.size(); i++) {
            for (int j = target; j >= nums[i]; j--) {
                dp[j] = max(dp[j], dp[j - nums[i]] + nums[i]);
            }
        }
        // 集合中的元素正好可以凑成总和target
        if (dp[target] == target)
            return true;
        return false;
    }
};